src/validation.cpp

```diff
@@ -5291,6 +5291,262 @@ bool ExpectedAssumeutxo(const int height, uint256& expected_out, const CChainPar
+
+    bool out_of_coins{false};
+    try {
+        coins_file >> outpoint;
+    } catch (const std::ios_base::failure&) {
+        // We expect an exception since we should be out of coins.
+        out_of_coins = true;
+    }
+    if (!out_of_coins) {
+        LogPrintf("[snapshot] bad snapshot - coins left over after deserializing %d coins\n",
+            coins_count - coins_left);
+        return false;
+    }
+
+    LogPrintf("[snapshot] loaded %d (%.2f MB) coins from snapshot %s\n",
+        coins_count - coins_left,
+        coins_cache.DynamicMemoryUsage() / (1000 * 1000),
+        base_blockhash.ToString());
+
+    LogPrintf("[snapshot] flushing snapshot chainstate to disk\n");
+    // No need to acquire cs_main since this chainstate isn't being used yet.
+    coins_cache.Flush(); // TODO: if #17487 is merged, add erase=false here for better performance.
+
+    assert(coins_cache.GetBestBlock() == base_blockhash);
+
+    CCoinsStats stats;
+    auto breakpoint_fnc = [] { /* TODO insert breakpoint here? */ };
+    if (!GetUTXOStats(&snapshot_chainstate.CoinsDB(), stats, CoinStatsHashType::HASH_SERIALIZED, breakpoint_fnc)) {
+        LogPrintf("[snapshot] failed to generate coins stats\n");
+        return false;
+    }
+
+    // Ensure that the base blockhash appears in the known chain of valid headers. We're willing to
+    // wait a bit here because the snapshot may have been loaded on startup, before we've
+    // received headers from the network.
+
+    int max_secs_to_wait_for_headers = 60 * 10;
+    CBlockIndex* snapshot_start_block = nullptr;
+
+    while (max_secs_to_wait_for_headers > 0) {
+        snapshot_start_block = LookupBlockIndex(base_blockhash);
+        max_secs_to_wait_for_headers -= 1;
+
+        if (!snapshot_start_block) {
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        } else {
+            break;
+        }
+    }
+
+    if (snapshot_start_block == nullptr) {
+        LogPrintf("[snapshot] timed out waiting for snapshot start blockheader %s\n",
+            base_blockhash.ToString());
+        return false;
+    }
+
+    // Assert that the deserialized chainstate contents match the expected assumeutxo value.
+
+    uint256 expected_contents_hash;
+    int base_height = snapshot_start_block->nHeight;
+
+    if (!ExpectedAssumeutxo(base_height, expected_contents_hash, ::Params())) {
+        LogPrintf("[snapshot] assumeutxo value in snapshot metadata not valid for " /* Continued */
+            "height %s - refusing to load snapshot\n", base_height);
+        return false;
+    }
+
+    if (stats.hashSerialized != expected_contents_hash) {
+        LogPrintf("[snapshot] bad snapshot content hash: expected %s, got %s\n",
+            expected_contents_hash.ToString(), stats.hashSerialized.ToString());
+        return false;
+    }
+
+    snapshot_chainstate.m_chain.SetTip(snapshot_start_block);
+
+    // Fake various pieces of CBlockIndex state:
```

Comment:
Ideally you're right here, but there's a lot of complexity in cleaning up the "consumer" side of this data. Take for example `nChainTx`; in order to avoid faking that value, there are various points of consumption in net_processing (by way of `HaveTxsDownloaded()`) that need to be addressed. What makes these sites complicated is that they lack the chainstate context, having access only to the CBlockIndex objects.

I'm not saying it isn't possible to do what you're suggesting here, but I think it will probably involve some deep thinking about how much regions like net_processing should know about chainstate semantics. I'm going to leave this for someone else to think about unless there is a fundamental objection to fudging this data for snapshot chainstates.

Reply:
Oh sorry, somehow misinterpreted what you'd said. Added some comments in `chain.h` that at least provide references to where the faked values are created and used; if you can think of anything else that'd be useful I'm happy to apply a patch.